File: v12_omniscience/scripts/psych_factors.py
import numpy as np
from datetime import datetime, timedelta

class PsychFactors:
    """
    Implements 'The Tilt Tensor' and 'Fatigue Decay' logic.
    """

    def calculate_tilt_tensor(self, last_game_margin, spread_diff, is_loss=True):
        """
        Quantifies a team's performance variance after a 'High-Stress/Low-Outcome' event.
        Tilt = (abs(Actual_Margin - Spread) / Time_Delta) * Impact_Weight
        """
        if not is_loss:
            return 0.0 # Positive wins rarely cause 'tilt' in the same negative sense
            
        # A 'heartbreaking loss' is one where they lost by more than the spread in a close game,
        # or lost a game they were heavily favored in.
        stress_score = abs(last_game_margin - spread_diff)
        
        # Normalize stress score to [0, 1]
        tilt_factor = min(1.0, stress_score / 20.0)
        return round(float(tilt_factor), 4)

    def calculate_fatigue_decay(self, minutes_played_7d):
        """
        Recovery-curve variable (R) based on minutes played in a rolling 7-day window.
        Fatigue = 1 - exp(-minutes / Baseline)
        """
        # Baseline: 120 minutes in 7 days is high for a star, 240 for a team average?
        # Let's say 200 minutes is the 'red zone'.
        baseline = 200.0
        r_value = 1 - np.exp(-minutes_played_7d / baseline)
        return round(float(r_value), 4)

if __name__ == "__main__":
    psych = PsychFactors()
    # Scenario: Team lost by 1 point at the buzzer when they were -5 favorites.
    tilt = psych.calculate_tilt_tensor(-1, -5)
    print(f"🧠 Tilt Tensor [Heartbreaker]: {tilt}")
    
    # Scenario: Star player played 180 minutes in last 7 days.
    fatigue = psych.calculate_fatigue_decay(180)
    print(f"💤 Fatigue Decay Score: {fatigue}")

File: src/services/psychFactors.ts
/**
 * Implements 'The Tilt Tensor' and 'Fatigue Decay' logic along with 'Swarm Consensus'.
 */

import { SwarmPredictor } from './swarmPredictor';
import type { SwarmSimulationParams, SwarmConsensus } from './swarmPredictor';

export class PsychFactors {
    /**
     * Quantifies a team's performance variance after a 'High-Stress/Low-Outcome' event.
     * Tilt = (abs(Actual_Margin - Spread) / Time_Delta) * Impact_Weight
     */
    static calculateTiltTensor(lastGameMargin: number, spreadDiff: number, isLoss: boolean = true): number {
        if (!isLoss) {
            return 0.0; // Positive wins rarely cause 'tilt' in the same negative sense
        }
        
        // A 'heartbreaking loss' is one where they lost by more than the spread in a close game,
        // or lost a game they were heavily favored in.
        const stressScore = Math.abs(lastGameMargin - spreadDiff);
        
        // Normalize stress score to [0, 1]
        const tiltFactor = Math.min(1.0, stressScore / 20.0);
        return parseFloat(tiltFactor.toFixed(4));
    }

    /**
     * Recovery-curve variable (R) based on minutes played in a rolling 7-day window.
     * Fatigue = 1 - exp(-minutes / Baseline)
     */
    static calculateFatigueDecay(minutesPlayed7d: number): number {
        // Baseline: 120 minutes in 7 days is high for a star, 240 for a team average?
        // Let's say 200 minutes is the 'red zone'.
        const baseline = 200.0;
        const rValue = 1 - Math.exp(-minutesPlayed7d / baseline);
        return parseFloat(rValue.toFixed(4));
    }

    /**
     * Integrates Swarm Intelligence by running a multi-agent debate simulation.
     * The resulting swarm confidence is combined with negative psych factors (Tilt, Fatigue)
     * to provide a final Confluence Multiplier.
     */
    static getSwarmConfluenceMultiplier(swarmParams: SwarmSimulationParams, tiltFactor: number, fatigueDecay: number): number {
        const consensus: SwarmConsensus = SwarmPredictor.runSimulation(swarmParams);
        
        // Base multiplier starts at Swarm predicted probability
        const baseMultiplier = consensus.predictedProbability;
        
        // If swarm confidence is very high, it mitigates some tilt/fatigue
        const confidenceBuffer = consensus.swarmConfidence * 0.5; // Up to 50% buffer

        // Effective negative factors
        const effectiveTilt = Math.max(0, tiltFactor - confidenceBuffer);
        const effectiveFatigue = Math.max(0, fatigueDecay - (confidenceBuffer / 2));

        // Penalty application
        const finalMultiplier = baseMultiplier * (1 - effectiveTilt) * (1 - effectiveFatigue);
        
        // Return bounded [0, 1]
        return parseFloat(Math.max(0, Math.min(1, finalMultiplier)).toFixed(4));
    }
}
